### PR TITLE
feat(api): Add user registration endpoint

### PR DESCRIPTION
**Pull request by dcramer**
Thie exposes /api/0/auth/register/ which mirrors the logic in user registration.

It also reconciles some duplicate behavior with rate limiting and metrics reporting throughout both registration and login flows.

Note: This duplicates the code in the RegistrationForm to leverage a DRF serializer, so that it can be extended and reused in getsentry. This should be fine as long as our intent is to move away from server-generated forms for the registration view in a near future.

### CODE REVIEW

File: src/sentry/api/endpoints/auth_login.py [hunk 1]
```diff
@@ -21,23 +19,6 @@ def post(self, request, organization=None, *args, **kwargs):
         """
         login_form = AuthenticationForm(request, request.data)
 
-        # Rate limit logins
-        is_limited = ratelimiter.is_limited(
```
Comment (dcramer) [inline src/sentry/api/endpoints/auth_login.py:25]: turns out the form was doing all of this..

File: src/sentry/api/endpoints/auth_register.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+ERR_UNKNOWN = "An unknown error occurred while creating the account."
+
+
+def send_confirm_emails(user):
+    for email in user.emails.filter(is_verified=False):
+        user.send_confirm_email_singular(email=email, is_new_user=True)
+
+
+class AuthRegisterSerializer(CamelSnakeSerializer):
```
Comment (dcramer) [inline src/sentry/api/endpoints/auth_register.py:29]: this does not implement the org creation logic, as its fairly complex, and its simpler for us to do that in our override in getsentry given we want to use a custom code path (provision)

File: src/sentry/web/forms/accounts.py [hunk 3]
```diff
@@ -107,7 +111,7 @@ def _is_user_rate_limited(self):
         if not username:
             return False
 
-        return ratelimiter.is_limited(f"auth:username:{username}", limit)
+        return ratelimiter.is_limited(f"auth:username:{md5_text(username)}", limit)
```
Comment (dcramer) [inline src/sentry/web/forms/accounts.py:114]: this was not uniform, but is safer to wrap in md5

File: src/sentry/web/frontend/auth_login.py [hunk 4]
```diff
@@ -163,27 +163,7 @@ def handle_basic_auth(self, request, **kwargs):
...
 
         elif request.method == "POST":
-            from sentry.app import ratelimiter
-            from sentry.utils.hashlib import md5_text
-
-            login_attempt = (
-                op == "login" and request.POST.get("username") and request.POST.get("password")
-            )
-
-            if login_attempt and ratelimiter.is_limited(
```
Comment (dcramer) [inline src/sentry/web/frontend/auth_login.py:173]: this rate limit was duplicated..

File: tests/sentry/api/endpoints/test_auth_register.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+class AuthRegisterEndpointTest(APITestCase):
+    path = "/api/0/auth/register/"
+
+    def setUp(self):
+        super().setUp()
+        options.set("auth.allow-registration", True)
+
+    @pytest.mark.xfail
+    def test_reactivate(self):
```
Comment (dcramer) [inline tests/sentry/api/endpoints/test_auth_register.py:18]: untested paths are marked xfail

**Comment by dcramer** [comment]
An alterantive approach - taken by the /auth/login/ endpoint - would be to pull in and instantiate the RegistrationForm vs creating a serializer here. That would require us to take a different a short-term approach in getsentry, but would prevent duplication of any of the registration logic within sentry itself. I'm personally ok with this, as it means we can mostly avoid touching the server-generated forms.

File: src/sentry/api/endpoints/auth_register.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        user = User(username=data["email"], email=data["email"], name=data["name"])
+        user.set_password(data["password"])
+        user.save()
+
+        # Authenticate user into the current request context
+        #
+        # XXX: This is a side-effect, which is maybe a little confusing as
+        #      creation of an account using this serializer will also
+        #      authenticate the created user.
+        user.backend = settings.AUTHENTICATION_BACKENDS[0]
```
Comment (dcramer) [inline src/sentry/api/endpoints/auth_register.py:79]: realistically we probably shouldn't be doing this given an API can be queried many ways.. its not obvious to me if this creates an attack vector (I dont think it does), but it at the very least would create session constructs where none is needed. that'd likely require other changes on our frontend though so for now its ok to keep this.

Comment (dcramer) [reply src/sentry/api/endpoints/auth_register.py:79]: auth_login has same issue - we also probably could change this to simply call `auth.login()` though that has some complexity that shouldn't be very relevant here

File: src/sentry/web/frontend/auth_login.py (hunk 4 above)
Comment (EvanPurkhiser) [reply src/sentry/web/frontend/auth_login.py:173]: nice

**Review by evanpurkhiser — APPROVED** [review]
Looks pretty reasonable to me

**Comment by dcramer** [comment]
see #26054

**Comment by dcramer** [comment]
I dont like the abuse vector this opens currently - so im going to merge in my recaptcha work and require recaptcha on this endpoint

**Comment by dcramer** [comment]
Updated to require CSRF on both login and register views, which mean they become unavailable outside of our UI.

**Comment by dcramer** [comment]
Ok this patch is scarier now from a scope, but should improve security quite a lot.

It now changes the SessionAuthentication backend, which is always the last thing we check (e.g. so a bearer token takes priority) to _always_ require CSRF, even if its an anonymous request. That assumption that its the last request is very important, as a normal API token would bypass CSRF.

This means a few views where you could have done a POST (or similar) request will no longer work without CSRF, such as POST /auth/. I'm not aware of this causing any issues with those, as those should generally be done via our web app, which passes CSRF along.

**Comment by dcramer** [comment]
I will look to merge this in the morning.

I did take another glance at the endpoints and I realized it should have no impact as all those endpoints already required authentication.

**Pull request merged**